pkg/kubelet/volumemanager/cache/desired_state_of_world.go

```diff
@@ -278,6 +298,9 @@ func (dsw *desiredStateOfWorld) DeletePodFromVolume(
 		// Delete volume if no child pods left
 		delete(dsw.volumesToMount, volumeName)
 	}
+
+	// create metric for volume unmount
+	dsw.unmountCompletionCallbackMap[volumeName] = metrics.VolumeManagerOperationCompleteHook(volumePluginName, "volume_unmount")
```

Comment:
This is not the right place to add metrics. This code is responsible for updating internal state. Which is one step of the operation. We should be adding metrics at the operation executor layer.

@davidz627 recently added common event logic and did a good job of factoring. His code may be worth a look. Will add him as a reviewer on this PR.

Reply:
Thanks for the feedback. I am not familiar with the event logic functionality. Will dig into it.

Reply:
You can take a look at https://github.com/kubernetes/kubernetes/pull/56872, it touches most of the places where events and metrics are being generated in the operation executor.

Specifically : https://github.com/kubernetes/kubernetes/pull/56872/files#diff-450e811a4953f760ff1594ede8b2037eR344 if you take a look at the "CompleteFunc's" they handle some sort of metric generation already, maybe you can augment those.

Reply:
This is bit tricky. @davidz627 the operation hooks you linked are used for tracking operation level metrics once operation has already started. 

This PR is for metrics that try to cover entirety of mount or umount operation (please refer to linked Google Doc for more details). Having a operation hook on umount on its own is not useful, we need a way to store that hook at the time when volume is removed from desired_state_of_world and we need to call the hook when umount actually finishes. The goal of these metrics is to measure - total time spent inside these controllers.

Reply:
The only other thing that I could/might do is wrap the callback in a volumesToUnmount struct in the event there is more information/state we need to carry forward for other operations then rename `unmountCompletionCallbackMap map[v1.UniqueVolumeName]func(error)` to something like `volumesToUnmount map[v1.UniqueVolumeName]volumeToUnmount`. I just started with the basic/simple case (just the callback itself) until we actually need to carry more state forward.

@gnufied @davidz627 @saad-ali  Please advise on how you want to move forward.

Reply:
Thanks for the explanation @gnufied. Now that I understand the use cases better, I understand why the code is structured this way. So next question: is it worth the added complexity in the code (and potential bugs; another data structure, `unmountCompletionCallbackMap`, that may get out of sync) to add this metric?